cpp/src/arrow/compute/kernels/vector_is_monotonic.cc

```diff
@@ -0,0 +1,321 @@
+}
+
+template <typename DataType>
+constexpr enable_if_not_floating_point<DataType, typename DataType::c_type> max() {
+  return std::numeric_limits<typename DataType::c_type>::max();
+}
+
+template <typename DataType>
+Status IsMonotonic(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+  using ArrayType = typename TypeTraits<DataType>::ArrayType;
+  using CType = typename TypeTraits<DataType>::CType;
+
+  auto options = IsMonotonicState::Get(ctx);
+
+  // Check batch size
+  if (batch.values.size() != 1) {
+    return Status::Invalid("IsMonotonic expects a single datum (array) as input");
+  }
+
+  // Safety:
+  // - Made sure there is at least one input datum.
+  Datum input = batch[0];
+
+  // Validate input datum type (useful for direct invocation only).
+  if (!input.is_array()) {
+    return Status::Invalid("IsMonotonic expects array datum as input");
+  }
+
+  // Safety:
+  // - Made sure that the input datum is an array.
+  const std::shared_ptr<ArrayData>& array_data = input.array();
+  ArrayType array(array_data);
+
+  // Return early if there are NaNs, zero elements or one element in the array.
+  // And return early if there are only nulls.
+  if (array.length() <= 1 || array.null_count() == array.length()) {
+    if (std::any_of(array.begin(), array.end(), isnan<DataType>)) {
+      return IsMonotonicOutput(false, false, false, false, out);
+    } else {
+      // It is strictly increasing if there are zero or one elements or when nulls are
+      // ignored.
+      bool strictly =
+          array.length() <= 1 ||
+          options.null_handling == IsMonotonicOptions::NullHandling::IGNORE_NULLS;
+      return IsMonotonicOutput(true, strictly, true, strictly, out);
+    }
+  }
+
+  // Set null value based on option.
+  const CType null_value =
+      options.null_handling == IsMonotonicOptions::NullHandling::USE_MIN_VALUE
+          ? min<DataType>()
+          : max<DataType>();
+
+  bool increasing = true, strictly_increasing = true, decreasing = true,
+       strictly_decreasing = true;
+
+  // Safety:
+  // - Made sure that the length is at least 2 above.
+  for (auto a = array.begin(), b = ++array.begin(); b != array.end();) {
+    auto current = *a;
+    auto next = *b;
+
+    // Handle nulls.
+    if (options.null_handling == IsMonotonicOptions::NullHandling::IGNORE_NULLS) {
+      // Forward both iterators to search for a non-null value. The loop exit
+      // condition prevents reading past the end.
+      if (!current.has_value()) {
+        ++a;
+        ++b;
+        continue;
+      }
+      // Once we have a value for current we should also make sure that next has a
+      // value. The loop exit condition prevents reading past the end.
+      if (!next.has_value()) {
+        ++b;
+        continue;
+      }
+    } else {
+      // Based on the function options set null values to min/max.
+      current = current.value_or(null_value);
+      next = next.value_or(null_value);
```

Comment:
Are these local copies being modified, correct? An Boolean function should not modify its inputs.